### DISCUSSION TITLE
Invoice: projects merged instead of per unit price / rate

### DISCUSSION THREAD

**Discussion by SaturnusDJ** [kimai/kimai#4180] [Questions & Answers]
The default template and also other templates like https://github.com/kimai/invoice-templates/blob/main/grouped-by-projects/grouped-by-projects.pdf.twig display a project multiple times for all the unit prices or rates that happen inside it.

Example:
Project1 has two members; User1 with an hourly rate of 30 and user2 with an hourly rate of 50.
The template would list project1 two times, for each user.

Trying to have a template where this is all merged and the project is only shown once. Tried various things with the default template while having the debug template on screen too. E.g. looking for if this merged number is available to select, but seems like it is not.

Any tips?

**Comment by kevinpapst** [comment]
Each invoice line has a single price a unit and a total price.
This does not work with mixed hourly rates. 
Therefor independent hourly rates are split into single invoice lines.
That has nothing to do with the template, it happens in the Calculator, before the template is rendered:
https://github.com/kimai/kimai/blob/main/src/Invoice/Calculator/AbstractSumInvoiceCalculator.php#L45

The only chance you have is to select the "No grouping" calculator, which gives you access to all items and then do the merging yourself.

**Reply by SaturnusDJ** [reply]
Ah I understand now. Is the calculator selected with "Grouping of invoice lines"?
I have selected "project" there.

![image](https://github.com/kimai/kimai/assets/14320018/8aa40d03-1b41-446c-9159-27988122c0f5)

**Reply by kevinpapst** [reply]
Yes, exactly. Choose "Default" and you can try grouping in your template.

**Reply by SaturnusDJ** [reply]
Tried like https://stackoverflow.com/questions/55299799/how-to-group-by-column-name-in-twig but the required plugin https://github.com/dpolac/twig-lambda does not support PHP 8 (yet).

Tried other things like https://stackoverflow.com/a/23607348/2908446. It is very difficult, I just cannot get it to work.

**Reply by kevinpapst** [reply]
You need to create an array with the key being the project id and value being the rate: 
```twig
{% set map = {} %}
{% for entry in entries %}
{% set key = 'p_' ~ entry.project.id %}
{% set rate = 0 %}
{% if map[key] is defined %}
{% set rate = map[key] %}
{% endif %}
{% set map = {(key): (rate + entry.rate)} %}
{% endfor %}
```
This is not complete and not tested! 
You might want to store more values (like project name, duration, amount...) in the `map`.

At the end, do a loop over the map and create your invoice table from that.

**Reply by SaturnusDJ** [reply]
Thanks!
Thing are happening but not the right ones haha.

I changed `entry.project.id` to `entry['entry.project']` to get the value, and left out id to get the project name. This works. Using `entry['entry.rate']` instead of `entry.rate` gives server error 500.


```
    {% set map = {} %}
    {% for entry in entries %}
      {% set key = entry['entry.project'] %}
      {% set rate = 0 %}
      {% if map[key] is defined %}
        {% set rate = map[key] %}
      {% endif %}
      {% set map = {(key): (rate + entry.rate)} %}
    {% endfor %}

        {% for key, rate in map %}
            <tr>
                <td>  {{key}}  </td>
                <td nowrap class="text-nowrap text-right">{{ rate }}</td>
            </tr>
        {% endfor %}
```
Like this returns only one project.

```
    {% set map = {} %}
    {% for entry in entries %}
      {% set key = entry['entry.project'] %}
      {% set rate = 0 %}
      {% if map[key] is defined %}
        {% set rate = map[key] %}
      {% endif %}
      {% set map = {(key): (rate + entry.rate)} %}
            <tr>
                <td>  {{key}}  </td>
                <td nowrap class="text-nowrap text-right">{{ rate }}</td>
            </tr>
        {% endfor %}
```
Like this gives all the projects but not grouped by project, so there are duplicates.
